spec/immutable-data-structures.html

```diff
@@ -6,27 +6,573 @@ <h1>Immutable Data Structures</h1>
+          1. Remove the first element of _list_.
+          1. Return a new Tuple value whose [[Sequence]] is _list_.
+        </emu-alg>
+
+      </emu-clause>
+      <emu-clause id="sec-tuple.prototype.sliced">
+        <h1>Tuple.prototype.sliced ( _start_, _end_ )</h1>
+        <p>When the `sliced` method is called with two arguments, _start_ and _end_, and returns a tuple containing the elements of the tuple from element _start_ up to, but not including, element _end_ (or through the end of the tuple if _end_ is *undefined*). If _start_ is negative, it is treated as <emu-eqn>_length_ + _start_</emu-eqn> where _length_ is the length of the tuple. If _end_ is negative, it is treated as <emu-eqn>_length_ + _end_</emu-eqn> where _length_ is the length of the tuple.</p>
+        <p>The following steps are taken:</p>
+        <emu-alg>
+          1. Let _T_ be ? thisTupleValue(*this* value).
+          1. Let _list_ be _T_'s [[Sequence]] value.
+          1. Let _len_ be the number of elements in _list_.
+          1. Let _relativeStart_ be ? ToInteger(_start_).
+          1. If _relativeStart_ &lt; 0, let _k_ be max((_len_ + _relativeStart_), 0); else let _k_ be min(_relativeStart_, _len_).
+          1. If _end_ is *undefined*, let _relativeEnd_ be _len_; else let _relativeEnd_ be ? ToInteger(_end_).
+          1. If _relativeEnd_ &lt; 0, let _final_ be max((_len_ + _relativeEnd_), 0); else let _final_ be min(_relativeEnd_, _len_).
+          1. Let _count_ be max(_final_ - _k_, 0).
+          1. Let _newList_ be a new empty List.
+          1. Let _n_ be 0.
+          1. Repeat, while _k_ &lt; _final_,
+            1. Let _kValue_ be _list_[_k_].
+            1. If Type(_kValue_) is Object, throw a *TypeError* exception.
+            1. Append _kValue_ to the end of _newList_.
+            1. Set _k_ to _k_ + 1.
+            1. Set _n_ to _n_ + 1.
+          1. Return a new Tuple value whose [[Sequence]] is _newList_.
+        </emu-alg>
+      </emu-clause>
+      <emu-clause id="sec-array.prototype.sort">
+        <h1>Tuple.prototype.sorted ( _comparefn_ )</h1>
+        <p>A new tuple the elements from the original tuple is returned, where the elements are sorted. The sort must be stable (that is, elements that compare equal must remain in their original order). If _comparefn_ is not *undefined*, it should be a function that accepts two arguments _x_ and _y_ and returns a negative value if _x_ &lt; _y_, zero if _x_ = _y_, or a positive value if _x_ &gt; _y_.</p>
+        <p>Upon entry, the following steps are performed to initialize evaluation of the `sorted` function:</p>
+        <emu-alg>
+          1. If _comparefn_ is not *undefined* and IsCallable(_comparefn_) is *false*, throw a *TypeError* exception.
+          1. Let _T_ be ? thisTupleValue(*this* value).
+          1. Let _list_ be a new List containing the elements of _T_'s [[Sequence]] value.
+          1. Let _len_ be the number of elements in _list_.
+        </emu-alg>

```

Comment:
It would be good if you could avoid duplicating the below section with Array.prototype.sort, since it's just the same, right?

Reply:
Most of it is the same, except with the handling of sparse arrays removed. We could share more here. What would be the best way to share more of this? Not sure how to handle it, since it isn't quite an abstract-op, very prose-y.